apps/dashboard/src/components/primitives/field-editor/variable-plugin/variable-pill-widget.ts

```diff
@@ -0,0 +1,79 @@
+import { WidgetType } from '@uiw/react-codemirror';
+import { MODIFIERS_CLASS, VARIABLE_PILL_CLASS } from './constants';
+
+export class VariablePillWidget extends WidgetType {
+  private clickHandler: (e: MouseEvent) => void;
+
+  constructor(
+    private variableName: string,
+    private fullVariableName: string,
+    private start: number,
+    private end: number,
+    private hasModifiers: boolean,
+    private onSelect?: (value: string, from: number, to: number) => void
+  ) {
+    super();
+    this.clickHandler = (e: MouseEvent) => {
+      e.preventDefault();
+      e.stopPropagation();
+
+      // setTimeout is used to defer the selection until after CodeMirror's own click handling
+      // This prevents race conditions where our selection might be immediately cleared by the editor
+      setTimeout(() => {
+        this.onSelect?.(this.fullVariableName, this.start, this.end);
+      }, 0);
+    };
+  }
+
+  toDOM() {
+    const span = document.createElement('span');
+    const pillClass = `${VARIABLE_PILL_CLASS} ${this.hasModifiers ? MODIFIERS_CLASS : ''}`;
+
+    span.className = pillClass;
+
+    // Stores the complete variable expression including any filters/transformers
+    span.setAttribute('data-variable', this.fullVariableName);
+
+    span.setAttribute('data-start', this.start.toString());
+    span.setAttribute('data-end', this.end.toString());
+
+    // Contains the clean variable name shown to the user
+    span.setAttribute('data-display', this.variableName);
+
+    span.textContent = this.variableName;
+
+    span.addEventListener('mousedown', this.clickHandler);
+
+    return span;
+  }
+
+  /**
+   * Determines if two VariablePillWidget instances are equal by comparing all their properties.
+   * Used by CodeMirror to optimize re-rendering.
+   */
+  eq(other: VariablePillWidget) {
+    return (
+      other.variableName === this.variableName &&
+      other.fullVariableName === this.fullVariableName &&
```

Comment:
The full expression of the variable along with the start and end position should be enough for the comparison.